File: python/leetcode/palindrome_number/palindrome_number.py
'''
https://leetcode.com/problems/palindrome-number/
Determine whether an integer is a palindrome
'''
class Solution(object):
	def isPalindrome(self, x):
		"""
		:type x: int
		:rtype: bool
		"""
		if x < 0:
			return False

		if x < 10:
			return True

		xStr = str(x)
		i, j = 0, len(xStr)-1
		while i < j:
			if xStr[i] != xStr[j]:
				return False
			i+= 1
			j-= 1

		return True


if __name__ == '__main__':
	s = Solution()
	assert s.isPalindrome(1) == True
	assert s.isPalindrome(-1) == False
	assert s.isPalindrome(-121) == False
	assert s.isPalindrome(121) == True
	assert s.isPalindrome(11) == True
	assert s.isPalindrome(10) == False

File: haskell/leetcode/palindrome_number/palindrome_number.hs
{-
https://leetcode.com/problems/palindrome-number/

Sample Input
3
12
121
-121

Sample Output
False
True
False
-}

is_palindrome :: Int -> Bool
is_palindrome n
	| (n < 0) = False
	| (n < 10) = True
	| otherwise  = (nStr == reverse nStr)
	where nStr = show n
	

-- Read n lines and process separately
getLinesAndProcess :: Int -> IO ()
getLinesAndProcess 0 = return ()
getLinesAndProcess n = do
	num <- readLn :: IO Int
	putStrLn $ show (is_palindrome num)
	getLinesAndProcess (n-1)

main = do
	n <- readLn :: IO Int
	getLinesAndProcess n


{-
 - sample run:
 - 
 - $ echo -e "3\n1\n-11\n121" | runghc leetcode/palindrome_number/palindrome_number.hs
 - True
 - False
 - True
-}
